MyCryptoHQ/MyCrypto
# Address Manager

### Description
Adds functionality related to an "Address Book" to MyCrypto:
* An "Address Book" tab after logging in where new <Address: Label> pairs can be entered.
![address-book-label](https://user-images.githubusercontent.com/10479826/39088543-dbba999c-4568-11e8-88b0-05fa6ec8aeb0.png)

* The `AccountInfo` component on the right side of the account now displays one of three states based on whether the user has entered a label for their own address:

1. No label, a button shows prompting the quick creation.

![add-label-account-info](https://user-images.githubusercontent.com/10479826/39088562-1d0a6bde-4569-11e8-837c-892322c321e2.png)

2. A label exists, shows an edit icon as well as the label.

![screen shot 2018-04-21 at 1 37 42 pm](https://user-images.githubusercontent.com/10479826/39088565-2f268faa-4569-11e8-9cbc-5e28db4ada44.png)

3. The label is being edited from `AccountInfo`

![screen shot 2018-04-21 at 1 38 10 pm](https://user-images.githubusercontent.com/10479826/39088567-3f56660c-4569-11e8-9458-8cd98463fa16.png)

* The `AddressField` component now displays a dropdown when the value contains two or more characters. This dropdown is navigable via arrow keys and will enter the address into the field on selection.

![screen shot 2018-04-21 at 1 40 10 pm](https://user-images.githubusercontent.com/10479826/39088593-ac0fb4a6-4569-11e8-8edf-af3250932f84.png)

* If the value in the `AddressField` component matches a label, the label for the input field will change.

![screen shot 2018-04-21 at 1 42 09 pm](https://user-images.githubusercontent.com/10479826/39088603-e81a7ae4-4569-11e8-8845-3ff22f3ecfd3.png)


* Addresses in the `DeterministicWalletModal` now display a label on top if one exists.

![deterministic-wallet-modal](https://user-images.githubusercontent.com/10479826/39088534-c125748a-4568-11e8-84f1-c316941f9b9a.png)

### Changes
* Created a new tab in `Tabs/SendTransaction`: `AddressBook`
* Created new components: AddressBookTable, AddressBookTableRow, AddressLabel
* Created an addition to `AddressFieldFactory` that acts as a dropdown
* Implemented a fully tested `addressBook` reducer that adds labels to both LocalStorage and the redux state
* Added new tested selectors to get pieces of `addressBook`

### Steps to Test
1. Log into an account
2. Navigate to the `Address Book` tab
3. Click the "+" button next to the Add new address/label row
* The `Add new address` field should be selected
4. Enter an invalid address
5. Click the "+" button again
* The `Add new address` field should be selected again
6. Enter a valid Ethereum address
7. Enter the label "Minni"
8. Press <ENTER>
* A new row should be added to the table with the entered Address/Minni pair.
9. Click on the Minni field, change it to Minnie, press <ENTER>
* The field should no longer be highlighted.
10. On the `AccountInfo` component, press the `Add Label` button.
11. In the input, enter `Mine` and press enter.
12. The label should now show `Mine`
13. The `AddressBook` table in the active tab should now have a new entry for `Mine`.
14. Press the `X` button next to `Minnie` label.
* It should be removed.
15. Navigate to the `Send Ether & Tokens` tab.
16. Enter `Mi` into the To field.
* A dropdown should appear showing `Mine` with the address.
17. Press the <DOWN ARROW> key, then the <ENTER> key.
18. The address for `Mine` should populate the field and the label should read (check) Mine
